### PR TITLE
Updates to tauID python tool

### PR DESCRIPTION
**Pull request by mbluj**
#### PR description:

This PR updates a tool to run and then embed tauIDs to pat::Tau (RecoTauTag/RecoTau/python/tools/runTauIdMVA.py). The update enables possibility to run the tool for more than one pat::Tau collection in one job through specifying name of input collection and postix added to created modules and sequences. The new functionality does not affect any official workflows.

It is considered to backport this PR to 106X release series to be used on top on UL miniAOD.

#### PR validation:

Unit and (limited) matrix tests successful.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mbluj** [comment]
It was found that phase-2 workflows are affected by a typo - it is fixed now.

**Comment by perrotta** [comment]
please test

**Comment by mbluj** [comment]
Sorry for recent late commit; I decided to add to a unit test with the tool being updated execution of the tools with features added in this PR.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mbluj** [comment]
Recent commit adapts the miniAOD setup-tool to properly use the updated tauID tool. Use of the updated tauID tool actually simplifies a bit a  setup of additional tauIDs in miniAOD.
It was explicitly tested with problematic phase-2 workflows as well as their Run-2 counterparts.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mbluj** [comment]
It looks that matrix jenkins tests did not started - could you please issue again a command triggering the tests?

**Comment by perrotta** [comment]
@cmsbuild please test

File: RecoTauTag/RecoTau/python/tools/runTauIdMVA.py [hunk 1]
```diff
@@ -152,19 +156,19 @@ def runTauID(self):
                 if self.debug: print ("runTauID: not is_above_cmssw_version(9, 4, 4). Will update the list of available in DB samples to access 2017v1")
                 self.loadMVA_WPs_run2_2017()
 
-            self.process.rerunDiscriminationByIsolationOldDMMVArun2017v1raw = patDiscriminationByIsolationMVArun2v1raw.clone(
-                PATTauProducer = cms.InputTag('slimmedTaus'),
+            setattr(self.process,"rerunDiscriminationByIsolationOldDMMVArun2017v1raw"+self.postfix,patDiscriminationByIsolationMVArun2v1raw.clone(
```
Comment (slava77) [inline RecoTauTag/RecoTau/python/tools/runTauIdMVA.py:159]:
```suggestion
           _rerunDiscriminationByIsolationOldDMMVArun2017v1raw = "rerunDiscriminationByIsolationOldDMMVArun2017v1raw"+self.postfix
                        setattr(self.process,_rerunDiscriminationByIsolationOldDMMVArun2017v1raw,patDiscriminationByIsolationMVArun2v1raw.clone(
```
and reuse it below.
This perhaps also gives an opportunity to shorten the name for better readability

File: RecoTauTag/RecoTau/python/tools/runTauIdMVA.py [hunk 2]
```diff
@@ -152,19 +156,19 @@ def runTauID(self):
...
                 loadMVAfromDB = cms.bool(True),
                 mvaName = cms.string("RecoTauTag_tauIdMVAIsoDBoldDMwLT2017v1"),#RecoTauTag_tauIdMVAIsoDBoldDMwLT2016v1 writeTauIdDiscrMVAs
                 mvaOpt = cms.string("DBoldDMwLTwGJ"),
                 verbosity = cms.int32(0)
-            )
+            ))
 
-            self.process.rerunDiscriminationByIsolationOldDMMVArun2017v1 = patDiscriminationByIsolationMVArun2v1.clone(
-                PATTauProducer = cms.InputTag('slimmedTaus'),
+            setattr(self.process,"rerunDiscriminationByIsolationOldDMMVArun2017v1"+self.postfix,patDiscriminationByIsolationMVArun2v1.clone(
```
Comment (slava77) [inline RecoTauTag/RecoTau/python/tools/runTauIdMVA.py:168]:
```suggestion
         _rerunDiscriminationByIsolationOldDMMVArun2017v1 = "rerunDiscriminationByIsolationOldDMMVArun2017v1"+self.postfix
                        setattr(self.process,_rerunDiscriminationByIsolationOldDMMVArun2017v1,patDiscriminationByIsolationMVArun2v1.clone(
```
and replace below

File: RecoTauTag/RecoTau/python/tools/runTauIdMVA.py [hunk 3]
```diff
@@ -126,8 +130,8 @@ def loadMVA_WPs_run2_2017(self):
             )
 
     def runTauID(self):
-        self.process.rerunMvaIsolationTask = cms.Task()
-        self.process.rerunMvaIsolationSequence = cms.Sequence()
+        setattr(self.process,"rerunMvaIsolationTask"+self.postfix,cms.Task())
+        setattr(self.process,"rerunMvaIsolationSequence"+self.postfix,cms.Sequence())
```
Comment (slava77) [inline RecoTauTag/RecoTau/python/tools/runTauIdMVA.py:134]:
```suggestion
        setattr(self.process,"rerunMvaIsolationTask"+self.postfix,cms.Task())
        _rerunMvaIsolationTask = getattr(self.process,"rerunMvaIsolationTask"+self.postfix)
```
and replace `getattr(self.process,"rerunMvaIsolationTask"+self.postfix)` with `_rerunMvaIsolationTask` below
for much less verbose and also faster access

File: RecoTauTag/RecoTau/python/tools/runTauIdMVA.py [hunk 4]
```diff
@@ -787,13 +801,13 @@ def runTauID(self):
                 mvaName_wGwGSF_EC = 'RecoTauTag_antiElectron'+antiElectronDiscrMVA6_version+'_gbr_wGwGSF_EC',
                 mvaName_woGwGSF_BL = 'RecoTauTag_antiElectron'+antiElectronDiscrMVA6_version+'_gbr_woGwGSF_BL',
                 mvaName_woGwGSF_EC = 'RecoTauTag_antiElectron'+antiElectronDiscrMVA6_version+'_gbr_woGwGSF_EC'
-            )
+            ))
```
Comment (slava77) [inline RecoTauTag/RecoTau/python/tools/runTauIdMVA.py:804]:
```suggestion
            ))
            _patTauDiscriminationByElectronRejectionMVA62018Raw = getattr(self.process,"patTauDiscriminationByElectronRejectionMVA62018Raw"+self.postfix)
```
for easier access later;
please apply this to everything that now has more than one `getattr`

**Review by slava77 — COMMENTED** [review]
there are multiple cases of the same repeated operations, which better be reduced by defining a temporary. I think that `getattr` is especially something to avoid.

**Comment by mbluj** [comment]
In recent commit: reduced number of `getattr`/`setattr` and a set of other improvements as requested in the review.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/33150	9d5f04a
- code changes are in line with the PR description and the follow up review; the changes are not changing the default processing
- jenkins tests pass and comparisons with the baseline show no (relevant) differences

**Comment by qliphy** [comment]
+1

**Pull request merged**